test/unit/cmap/message_stream.test.js

```diff
@@ -18,117 +20,101 @@ function bufferToStream(buffer) {
-          'hex'
-        ),
-        documents: [{ $db: 'admin', [LEGACY_HELLO_COMMAND]: 1 }]
-      },
-
-      {
-        description: 'Invalid message size (negative)',
-        data: Buffer.from('ffffffff', 'hex'),
-        error: 'Invalid message size: -1'
-      },
-      {
-        description: 'Invalid message size (exceeds maximum)',
-        data: Buffer.from('01000004', 'hex'),
-        error: 'Invalid message size: 67108865, max allowed: 67108864'
-      }
-    ].forEach(test => {
-      it(test.description, function (done) {
-        const error = test.error;
-        const expectedMessageCount = test.expectedMessageCount || 1;
-        const inputStream = bufferToStream(test.data);
+describe('MessageStream', function () {
+  context('when the stream uses the streaming protocol', function () {
+    const response = { isWritablePrimary: true };
+    let firstHello;
+    let secondHello;
+    let thirdHello;
+
+    beforeEach(function () {
+      firstHello = generateOpMsgBuffer(response);
+      secondHello = generateOpMsgBuffer(response);
+      thirdHello = generateOpMsgBuffer(response);
+    });
+
+    it('only reads the last message in the buffer', async function () {
+      const inputStream = bufferToStream(Buffer.concat([firstHello, secondHello, thirdHello]));
+      const messageStream = new MessageStream();
+      messageStream.isStreamingProtocol = true;
+
+      inputStream.pipe(messageStream);
+      const messages = await once(messageStream, 'message');
+      const msg = messages[0];
+      msg.parse();
+      expect(msg).to.have.property('documents').that.deep.equals([response]);
+      // Make sure there is nothing left in the buffer.
+      expect(messageStream.buffer.length).to.equal(0);
+    });
+  });
+
+  context('when the stream is not using the streaming protocol', function () {
+    context('when the messages are valid', function () {
+      const response = { isWritablePrimary: true };
+      let firstHello;
+      let secondHello;
+      let thirdHello;
+      let messageCount = 0;
+
+      beforeEach(function () {
+        firstHello = generateOpMsgBuffer(response);
+        secondHello = generateOpMsgBuffer(response);
+        thirdHello = generateOpMsgBuffer(response);
+      });
+
+      it('reads all messages in the buffer', function (done) {
+        const inputStream = bufferToStream(Buffer.concat([firstHello, secondHello, thirdHello]));
         const messageStream = new MessageStream();
 
-        let messageCount = 0;
         messageStream.on('message', msg => {
           messageCount++;
-          if (error) {
-            done(new Error(`expected error: ${error}`));
-            return;
+          msg.parse();
+          expect(msg).to.have.property('documents').that.deep.equals([response]);
+          // Test will not complete until 3 messages processed.
+          if (messageCount === 3) {
+            done();
```

Comment:
you can actually use `on` as an async iterator (and you can probably use `once` for the error listeners) to make all these tests async/await: https://nodejs.org/api/events.html#eventsonemitter-eventname-options

Reply:
Updated to make all tests use async/await